proposals/Scoped-Custom-Element-Registries.md

```diff
@@ -0,0 +1,227 @@
+2. **Internal implementation elements:** A component may have private internal elements that it needs to register, but would prefer not to leak to the external environment.
+
+### Why use shadow roots as registration scopes?
+
+Shadow roots provide an encapsulation mechanism for isolating a DOM tree's, nodes, styles, and events from other scopes. The goal is to allow scopes to function without required coordination with other scopes. The globally shared custom element registry, however, requires coordination so that multiple scopes on a page agree on the registrations. Rather than invent a new scoping primitive to the DOM, it is natural to extend the shadow root scoping responsibilities to include custom element registrations.
+
+This approach also allows for a nice API by extending the DocumentOrShadowRoot interface and `ShadowRoot#attachShadow()`.
+
+## Overview
+
+This proposal allows user code to create new instances of `CustomElementRegistry`:
+
+```js
+const registry = new CustomElementRegistry();
+
+and associate them with a ShadowRoot:
+
+export class MyElement extends HTMLElement {
+  constructor() {
+    this.attachShadow({mode: 'open', registry});
+  }
+}
+```
+
+Such scoped registries can be populated with element definitions, completely under the control of the registry owner:
+
+```js
+import {OtherElement} from './my-element.js';
+
+registry.define('other-element', OtherElement);
+```
+
+Definitions in this registry do not apply to the main document, and vice-versa. The registry must contain definitions for all elements used.
+
+Once a registry and scope are created, element creation associated with the scope will use that registry to look up custom element definitions:
+
+```js
+this.shadowRoot.innerHTML = `<other-element></other-element>`;
+```
+
+These scoped registries will allow for different parts of a page to contain definitions for the same tag name.
+
+### Creating and using a CustomElementRegistry
+
+A new CustomElementRegistry is created with the CustomElementRegistry constructor, and attached to a ShadowRoot with the `registry` option to `HTMLElement.prototype.attachShadow`:
+
+```js
+import {OtherElement} from './my-element.js';
+
+const registry = new CustomElementRegistry();
+registry.define('other-element', OtherElement);
+
+export class MyElement extends HTMLElement {
+  constructor() {
+    this.attachShadow({mode: 'open', registry});
+  }
+}
+```
+
+### Scoped element creation APIs
+
+Element creation APIs, like `createElement()` and `innerHTML` can be grouped into global API (those on Document or Window) and scoped APIs (those on HTMLElement and ShadowRoot). The scoped APIs have an associated Node that can be used to look up a CustomElementRegistry and thus a scoped definition.
+
+In order to support scoped registries we add new scoped APIs, that were previously only available on `Document`, to `ShadowRoot`:
+
+* `createElement()`
+* `createElementNS()`
+* `importNode()`
```

Comment:
what about `adoptNode`?

Reply:
`adoptNode` doesn't create nodes or run upgrade steps (https://dom.spec.whatwg.org/#concept-node-adopt) so it doesn't need to find a registry.